File: advent_of_code/2024/python/03.py

import re

def part1(text):
    t = 0
    for expr in re.findall(
        r'mul\(\d+,\d+\)', text
    ):
        a, b = re.findall('\d+', expr)
        t += int(a) * int(b)
    return t

def part2(text):
    t = 0
    do = True
    for expr in re.findall(
        r'mul\(\d+,\d+\)|do\(\)|don\'t\(\)', text
    ):
        if expr == 'do()':
            do = True 
        elif expr == 'don\'t()':
            do = False
        else:
            if do:
                a, b = re.findall('\d+', expr)
                t += int(a) * int(b)
    return t

File: advent_of_code/2024/rust/src/03.rs
use regex::Regex;

fn part1(text: &str) -> i32 {
    let re = Regex::new(r"mul\(\d+,\d+\)").unwrap();
    let digit_re = Regex::new(r"\d+").unwrap();
    let mut total = 0;

    for expr in re.find_iter(text) {
        let numbers: Vec<i32> = digit_re
            .find_iter(expr.as_str())
            .map(|num| num.as_str().parse::<i32>().unwrap())
            .collect();
        if let [left, right] = numbers.as_slice() {
            total += left * right;
        }
    }

    total
}

fn part2(text: &str) -> i32 {
    let re = Regex::new(r"mul\(\d+,\d+\)|do\(\)|don't\(\)").unwrap();
    let digit_re = Regex::new(r"\d+").unwrap();
    let mut total = 0;
    let mut enabled = true;

    for expr in re.find_iter(text) {
        match expr.as_str() {
            "do()" => enabled = true,
            "don't()" => enabled = false,
            _ => {
                if enabled {
                    let numbers: Vec<i32> = digit_re
                        .find_iter(expr.as_str())
                        .map(|num| num.as_str().parse::<i32>().unwrap())
                        .collect();
                    if let [left, right] = numbers.as_slice() {
                        total += left * right;
                    }
                }
            }
        }
    }

    total
}

fn main() {
    aoc2024::solve(part1, part2);
}
